gui.js

```diff
@@ -3560,6 +3565,103 @@ var CGUI = function()
 
     return true;
   };
+  
+  // This function tries to automatically fill the currently
+  // selected range in the sequencer. If there's only one value
+  // at top, it repeats it. If there's two, it tries to create
+  // an increasing or decreasing sequence e.g. [1 2 0 0 0 0]
+  // becomes [1 2 3 4 5 6]. Three or more patterns are looped e.g.
+  // [1 2 1 2 0 0 0 0] becomes [1 2 1 2 1 2 1 2]
+  var fillSequenceRange = function()
+  {   
+    if (mSeqRow2 <= mSeqRow)
+      return;
+    // checks if the condition holds for any value between first and last, inclusive
+    // find the lastrow that is not all empty
+    outer:
+    for (var lastRow = mSeqRow2; lastRow > mSeqRow; --lastRow) {
+      for (var col = mSeqCol;col <= mSeqCol2;++col)
+        if (mSong.songData[col].p[lastRow]>0)
+          break outer;
+    }   
+    if (lastRow - mSeqRow >= 2) {     
+      // we have 3 or more rows of patterns, so loop them
+      var loopLength = lastRow-mSeqRow+1;   
+      var fillFunction = function(col,row) {
+        return mSong.songData[col].p[mSeqRow+(row-mSeqRow)%loopLength];
+      };    
+    } else {
+      // we have zero to two rows, try to repeate or extrapolate values
+      var fillFunction = function(col,row) {
+        var start = mSong.songData[col].p[mSeqRow];
+        if (!start) // no change if the value at starting row is empty
+          return mSong.songData[col].p[row];
+        var next = mSong.songData[col].p[mSeqRow+1];
+        var delta = !next ? 0 : next - start;       
+        var extrapolated = delta * (row-mSeqRow)+start;
+        return Math.max(Math.min(extrapolated,MAX_PATTERNS),1);
+      }        
+    }
+    for (var col = mSeqCol;col <= mSeqCol2;col++)
+      for (var row = mSeqRow;row <= mSeqRow2;row++)
+        mSong.songData[col].p[row] = fillFunction(col,row);     
+    updateSequencer();
+  }
+  
+  // This function interpolates the effect parameters
+  // between two values. The two commands have to be the same
+  // and the cells between them empty.
+  var fillFxRange = function()
+  {   
+    if (mFxTrackRow2 <= mFxTrackRow+2)
+      return; 
+    var pat = mSong.songData[mSeqCol].p[mSeqRow]-1;
+    if (pat < 0) 
+      return;
+    var f = mSong.songData[mSeqCol].c[pat].f;
+    var cmd1 = f[mFxTrackRow];
+    var cmd2 = f[mFxTrackRow2];
+    if (cmd1 != cmd2)
+      return;           
+    for (var row = mFxTrackRow+1; row < mFxTrackRow2; ++row)
+      if (f[row])
+        return;
+    var val1 = f[mFxTrackRow + mSong.patternLen];
+    var val2 = f[mFxTrackRow2 + mSong.patternLen];        
+    for (var row = mFxTrackRow+1; row < mFxTrackRow2; ++row)
+    {
+      f[row] = cmd1;
+      var alpha = (row - mFxTrackRow)/(mFxTrackRow2- mFxTrackRow+1)*1.0;
+      var roundedValue = ((alpha * val2 + (1 - alpha) * val1)+.5)|0;
```

Comment:
This interpolation may not be optional for every effect (e.g. arpeggio uses special encoding in the FX column, and some FX should probably use non-linear interpolation). That said, I think this solution is fine for now.